File: backend/services/order_service.py
from typing import Dict, Any, List
from dnse_client import DNSEClient
from core.data_structures import OrderRequest

class OrderService:
    def __init__(self):
        self.client = DNSEClient()
    
    def place_order(self, order_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Place a new order
        """
        order_request = OrderRequest(**order_data)
        return self.client.place_order(order_request)
    
    def get_order_details(self, order_id: str, account_no: str) -> Dict[str, Any]:
        """
        Get details of a specific order
        """
        return self.client.get_order_details(order_id, account_no)
    
    def cancel_order(self, order_id: str, account_no: str) -> Dict[str, Any]:
        """
        Cancel a pending order
        """
        return self.client.cancel_order(order_id, account_no)
    
    def get_pending_orders(self, account_no: str) -> List[Dict[str, Any]]:
        """
        Get all pending orders for an account
        """
        orders = self.client.get_orders(account_no)
        
        # Filter for pending orders only
        pending_statuses = ["pending", "pendingNew", "new", "partiallyFilled"]
        pending_orders = [order for order in orders if order.get("orderStatus") in pending_statuses]
        
        return pending_orders
        
    def get_orders(self, account_no: str) -> List[Dict[str, Any]]:
        """
        Get all orders for an account (includes orders with any status)
        
        Args:
            account_no: The account number
            
        Returns:
            List of all orders for the account
        """
        return self.client.get_orders(account_no)

File: ui/src/services/OrderService.js
// OrderService.js - API client for order-related endpoints

export default class OrderService {
  static async getPendingOrders(accountNo) {
    try {
      const response = await fetch(
        `/api/orders/pending?accountNo=${encodeURIComponent(accountNo)}`
      );
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || "Failed to fetch pending orders");
      }
      return await response.json();
    } catch (error) {
      console.error("Error fetching pending orders:", error);
      throw error;
    }
  }

  static async getOrderDetails(orderId, accountNo) {
    try {
      const response = await fetch(
        `/api/orders/${encodeURIComponent(
          orderId
        )}?accountNo=${encodeURIComponent(accountNo)}`
      );
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || "Failed to fetch order details");
      }
      return await response.json();
    } catch (error) {
      console.error("Error fetching order details:", error);
      throw error;
    }
  }

  static async cancelOrder(orderId, accountNo) {
    try {
      const response = await fetch(
        `/api/orders/${encodeURIComponent(
          orderId
        )}?accountNo=${encodeURIComponent(accountNo)}`,
        {
          method: "DELETE",
        }
      );
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || "Failed to cancel order");
      }
      return await response.json();
    } catch (error) {
      console.error("Error canceling order:", error);
      throw error;
    }
  }

  static async placeOrder(orderData) {
    try {
      const response = await fetch("/api/orders", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(orderData),
      });
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || "Failed to place order");
      }
      return await response.json();
    } catch (error) {
      console.error("Error placing order:", error);
      throw error;
    }
  }

  static async getOrderHistory(accountNo, filters = {}) {
    // Since there's no specific history API in DNSE, we'll use the regular orders endpoint
    // and filter on the client side
    const queryParams = new URLSearchParams();
    queryParams.append("accountNo", accountNo);

    // Add filters for backend filtering (status, symbol)
    if (filters.symbol) queryParams.append("symbol", filters.symbol);
    if (filters.status) queryParams.append("status", filters.status);

    try {
      // Use the orders/history endpoint which will use the main orders endpoint internally
      const response = await fetch(
        `/api/orders/history?${queryParams.toString()}`
      );
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || "Failed to fetch order history");
      }

      const orders = await response.json();

      // Additional client-side filtering for date range if needed
      let filteredOrders = orders;

      if (filters.startDate || filters.endDate) {
        filteredOrders = orders.filter((order) => {
          const orderDate = new Date(order.createdDate || order.transDate);

          if (filters.startDate) {
            const startDate = new Date(filters.startDate);
            if (orderDate < startDate) return false;
          }

          if (filters.endDate) {
            const endDate = new Date(filters.endDate);
            // Add a day to include the end date fully
            endDate.setDate(endDate.getDate() + 1);
            if (orderDate > endDate) return false;
          }

          return true;
        });
      }

      return filteredOrders;
    } catch (error) {
      console.error("Error fetching order history:", error);
      throw error;
    }
  }
}
